File: clevertap-core/src/main/java/com/clevertap/android/sdk/db/BaseDatabaseManager.java
package com.clevertap.android.sdk.db;

import android.content.Context;
import androidx.annotation.RestrictTo;
import androidx.annotation.RestrictTo.Scope;
import com.clevertap.android.sdk.events.EventGroup;
import org.json.JSONObject;

@RestrictTo(Scope.LIBRARY_GROUP)
public abstract class BaseDatabaseManager {

    public abstract void clearQueues(final Context context);

    public abstract QueueCursor getQueuedEvents(final Context context, final int batchSize,
            final QueueCursor previousCursor,
            final EventGroup eventGroup);

    public abstract void queueEventToDB(final Context context, final JSONObject event, final int type);

    abstract QueueCursor updateCursorForDBObject(JSONObject dbObject, QueueCursor cursor);

    public abstract DBAdapter loadDBAdapter(Context context);

    abstract QueueCursor getQueueCursor(final Context context, DBAdapter.Table table, final int batchSize,
            final QueueCursor previousCursor);

    abstract QueueCursor getQueuedDBEvents(final Context context, final int batchSize,
            final QueueCursor previousCursor);

    abstract QueueCursor getPushNotificationViewedQueuedEvents(final Context context, final int batchSize,
            final QueueCursor previousCursor);

    public abstract void queuePushNotificationViewedEventToDB(final Context context, final JSONObject event);

}

File: clevertap-core/src/main/java/com/clevertap/android/sdk/db/BaseDatabaseManager.kt
package com.clevertap.android.sdk.db

import android.content.Context
import com.clevertap.android.sdk.events.EventGroup
import org.json.JSONObject

internal interface BaseDatabaseManager {

    fun loadDBAdapter(context: Context): DBAdapter

    fun clearQueues(context: Context)

    /**
     * Gets queued events from the database
     * @param context Android context
     * @param batchSize Number of events to fetch (typically 50)
     * @param eventGroup Type of events to fetch
     * @return QueueData containing events and their IDs for cleanup
     */
    fun getQueuedEvents(
        context: Context,
        batchSize: Int,
        eventGroup: EventGroup
    ): QueueData

    fun queueEventToDB(context: Context, event: JSONObject, type: Int)

    fun queuePushNotificationViewedEventToDB(context: Context, event: JSONObject)

    fun getPushNotificationViewedQueuedEvents(context: Context, batchSize: Int): QueueData

    /**
     * Cleans up successfully sent events from the database
     * @param context Android context
     * @param eventIds List of event IDs to clean up from events table
     * @param profileEventIds List of event IDs to clean up from profileEvents table
     * @return true if cleanup successful, false otherwise
     */
    fun cleanupSentEvents(
        context: Context,
        eventIds: List<String>,
        profileEventIds: List<String>
    ) : Boolean

    fun cleanupPushNotificationEvents(
        context: Context,
        ids: List<String>
    ) : Boolean
}
